### ISSUE TITLE
`kube down` should be able to remove "external" storage containers

### ISSUE THREAD

**Issue by vrothberg** [containers/podman#20025]
### Discussed in https://github.com/containers/podman/discussions/20021

<div type='discussions-op-text'>

<sup>Originally posted by **poperigby** September 18, 2023</sup>
I have a collection of containers that are orchestrated with `.kube` files. Seemingly at random, a service will fail to start on reboot. When I go to start it manually with `podman kube play service.yaml`, it will say something like this:

```
Error: creating container storage: the container name "authelia-pod-authelia" is already in use by b8bf71cb1c5029f2dfee4f48e2be50c4bdeff7af9c99d5cd09d48b25eb8cbe7d. You have to remove that container to be able to reuse that name: that name is already in use
```

I then have to run something like this to fix it:

```
podman rm --storage b8bf71cb1c5029f2dfee4f48e2be50c4bdeff7af9c99d5cd09d48b25eb8cbe7d
```

It just recently (about a month ago), started doing this, so I'm not sure what happened.

Here's one of my `.kube` files:

```
[Install]
WantedBy=default.target

[Unit]
Requires=caddy.service
After=caddy.service

[Service]
TimeoutSec=900

[Kube]
Yaml=services/authelia/service.yaml
Network=reverse-proxy.network
```

Here's one of my Kubernetes style YAML files:

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: authelia-pod
  name: authelia-pod
spec:
  hostname: authelia-pod
  securityContext:
    seLinuxOptions:
      type: spc_t
  containers:
  - name: authelia
    image: ghcr.io/authelia/authelia:latest
    imagePullPolicy: Never
    #args:
    #- --config
    #- /config/configuration.yml
    env:
    - name: AUTHELIA_JWT_SECRET
      value: REDACTED
    - name: AUTHELIA_STORAGE_ENCRYPTION_KEY
      value: REDACTED
    - name: AUTHELIA_NOTIFIER_SMTP_PASSWORD
      value: REDACTED
    - name: AUTHELIA_IDENTITY_PROVIDERS_OIDC_HMAC_SECRET
      value: REDACTED
    - name: AUTHELIA_IDENTITY_PROVIDERS_OIDC_ISSUER_PRIVATE_KEY
      value: REDACTED
    - name: AUTHELIA_SESSION_SECRET
      value: REDACTED
    volumeMounts:
    - mountPath: /config
      name: authelia-config
  - name: redis
    image: docker.io/library/redis:latest
    #args:
    #- redis-server
    volumeMounts:
    - mountPath: /data
      name: redis-data
  volumes:
  - name: redis-data
    persistentVolumeClaim:
      claimName: redis
  - hostPath:
      path: /home/cassidy/.config/containers/systemd/services/authelia/authelia
      type: Directory
    name: authelia-config
```</div>

**Comment by vrothberg** [comment]
Copying my other [comment](https://github.com/containers/podman/discussions/20021#discussioncomment-7042787) from the discussion:
> The generated systemd units have a ExecStart=/usr/bin/podman kube play --replace (note the --replace). So to me it looks like the unit got shot down on stop (i.e., ExecStopPost=/usr/bin/podman kube down).
> 
> Without a reproducer we cannot be sure but I think it makes sense to have kube down and --replace remove "external" storage containers to account for such cases.

**Comment by Luap99** [comment]
Note this isn't strictly related to kube, `podman run --replace` would be another case where this might be important.
With 4.6 I saw several  instances of storage containers leaking, some even unable to remove unless you manually mount something in the storage dir (https://github.com/containers/podman/issues/19913#issuecomment-1717653277)

Most likely this is the result of increased error reporting in c/storage: https://github.com/containers/podman/issues/18831#issuecomment-1695792515
I would expect the latest fix (not in v4.6) to help in many cases.

Regardless of the cause I think removing the storage container by default makes sense, otherwise we could break all these non interactive systemd units which is very bad (and very much not self healing if a users has to fix this by hand).

**Comment by poperigby** [comment]
Bump

**Comment by rhatdan** [comment]
@umohnani8 PTAL

**Comment by rhatdan** [comment]
@vrothberg Did you ever fix this?

**Comment by vrothberg** [comment]
@rhatdan no, I only filed the issue

**Comment by tmzullinger** [comment]
> Note this isn't strictly related to kube, `podman run --replace` would be another case where this might be important. With 4.6 I saw several instances of storage containers leaking, some even unable to remove unless you manually mount something in the storage dir ([#19913 (comment)](https://github.com/containers/podman/issues/19913#issuecomment-1717653277))
> 
> Most likely this is the result of increased error reporting in c/storage: [#18831 (comment)](https://github.com/containers/podman/issues/18831#issuecomment-1695792515) I would expect the latest fix (not in v4.6) to help in many cases.
> 
> Regardless of the cause I think removing the storage container by default makes sense, otherwise we could break all these non interactive systemd units which is very bad (and very much not self healing if a users has to fix this by hand).

Agreed.  I have systemd generated units which call `podman pod` `start | stop` which recently failed to start after a reboot.  I had to manually rm the storage containers.  There is no `--replace` option for `podman pod` AFAICT, so I'm not sure how that should be fixed.
